src/mon/Paxos.cc

```diff
@@ -1414,53 +1414,57 @@ void Paxos::dispatch(MonOpRequestRef op)
 {
   ceph_assert(op->is_type_paxos());
   op->mark_paxos_event("dispatch");
-  PaxosServiceMessage *m = static_cast<PaxosServiceMessage*>(op->get_req());
+
+  if (op->get_req()->get_type() != MSG_MON_PAXOS) {
+    dout(0) << "Got unexpected message type " << op->get_req()->get_type()
+	    << " in Paxos::dispatch, aborting!" << dendl;
+    ceph_abort();
+  }
+  
+  MMonPaxos *pm = reinterpret_cast<MMonPaxos*>(op->get_req());
+
   // election in progress?
   if (!mon->is_leader() && !mon->is_peon()) {
-    dout(5) << "election in progress, dropping " << *m << dendl;
+    dout(5) << "election in progress, dropping " << *pm << dendl;
     return;    
   }
 
+
+
+  if (pm->epoch < mon->get_epoch()) {
```

Comment:
we drop the message on floor if `pm->epoch != get_epoch()` in `Monitor::dispatch_op()`.  see
https://github.com/ceph/ceph/blob/790472df84a9fd4c54630bc71abead5957aac371/src/mon/Monitor.cc#L4594-L4601

so i think we can just dispense with the check here.

Reply:
Oh crap. The assert failure is more complicated than I first thought or this epoch check is breaking somehow. Will need to check the logs more carefully. :(

Reply:
I don't think the problem is the epoch, but the leader check.

Looking at the logs, seems like for (e.g.) `mon.c`, around `04:59:43.571`, an election is finished and the leader declared `mon1`, with a quorum of `1,2,3,4,5,6,7,8` (i.e., excluding rank 0, `mon.f`). And that's the last election, not only this peon but others as well, go through.

Then we get to the assert: the cause is a `paxos(begin lc 3125)` sent by `mon.0` (i.e., `mon.f`); this monitor is not the leader and triggers the leader comparison of the assert.

Also, that paxos begin message is sent around `04:r9:48.096`, with `mon.f@0` still believing it's the leader, after having been SIGSTOP for about 16 seconds.

I think the real check we need to make is whether the incoming message is from the leader (right at the top of dispatch), and drop it if not.

Reply:
> I think the real check we need to make is whether the incoming message is from the leader (right at the top of dispatch), and drop it if not.

Actually, no. What needs to be checked is `if (mon->is_peon() && !(m->get_source().num() == mon->get_leader()) { return; }` after the election check.

Reply:
I know it's possible for two monitors to think they're leaders in the same epoch, but the circumstances are a bit narrow and I didn't think it could make it through the Elector processes to start delivering actual Paxos messages. I need to track down what happened; it's possible I managed to break something or else that there's a deeper issue than this.